lib/api/objectPut.js

```diff
@@ -91,77 +90,68 @@ function _storeIt(bucket, objectKey, objMD, authInfo, canonicalID, cipherBundle,
-                                bucket: bucketName,
-                                newByteLength: size,
-                                oldByteLength: prevContentLen,
-                            });
-                            return callback(null, contentMD5);
-                        });
-            });
-    }
-    log.trace('content-length is 0 so only storing metadata', {
-        method: 'services.metadataValidateAuthorization',
-    });
-    metadataStoreParams.contentMD5 = constants.emptyFileMd5;
-    const dataGetInfo = null;
-    return _storeInMDandDeleteData(bucketName, dataGetInfo, cipherBundle,
-        metadataStoreParams, dataToDelete,
-            logger.newRequestLoggerFromSerializedUids(log
-            .getSerializedUids()), (err, contentMD5) => {
-                if (err) {
-                    return callback(err);
-                }
-                pushMetric('putObject', log, {
-                    authInfo,
-                    bucket: bucketName,
-                    newByteLength: size,
-                    oldByteLength: prevContentLen,
+    // null - new object
+    // 0 or > 0 - existing object with content-length 0 or greater than 0
+    const requestLogger =
+        logger.newRequestLoggerFromSerializedUids(log.getSerializedUids());
+    return async.waterfall([
+        function storeData(next) {
+            if (size === 0) {
+                metadataStoreParams.contentMD5 = constants.emptyFileMd5;
+                return next(null, null, null);
+            }
+            return dataStore(objectKeyContext, cipherBundle, request, size,
+                    streamingV4Params, backendInfo, log, next);
+        },
+        function processDataResult(dataGetInfo, calculatedHash, next) {
+            if (dataGetInfo === null || dataGetInfo === undefined) {
+                return next(null, null);
+            }
+            // So that data retrieval information for MPU's and
+            // regular puts are stored in the same data structure,
+            // place the retrieval info here into a single element array
+            const { key, dataStoreName } = dataGetInfo;
+            const dataGetInfoArr = [{ key, size, start: 0, dataStoreName }];
+            if (cipherBundle) {
+                dataGetInfoArr[0].cryptoScheme = cipherBundle.cryptoScheme;
+                dataGetInfoArr[0].cipheredDataKey =
+                    cipherBundle.cipheredDataKey;
+            }
+            metadataStoreParams.contentMD5 = calculatedHash;
+            return next(null, dataGetInfoArr);
+        },
+        function getVersioningInfo(infoArr, next) {
+            dataGetInfoArr = infoArr;
+            return services.versioningPreprocessing(bucketName, bucketMD,
+                metadataStoreParams.objectKey, objMD, reqVersionId, log, next);
+        },
+        function createDeleteArray(options, next) {
+            if (!options.deleteNullVersionData) {
+                return next(null, options);
+            }
```

Comment:
OK
Actually when we delete the null version data is a little more complicated, so hope the comment helps explain: https://github.com/scality/S3/pull/651/commits/7456e668e263341e95e905bed998ced16ac8c78a#diff-e45defa8c8c53634533cf165cc9232ceR123

Normally as long as the null version is the latest version, it can be deleted just by using the master version's object md location property. We only have to get the null version's location when there is a null version that is not the latest version (can happen when we store the null version data when re-enabling versioning or enabling it for the first time).